app/modules/wiegand.c

```diff
@@ -0,0 +1,270 @@
+
+	default:
+		return originalKeyPress;
+	}
+}
+
+static bool wiegand_store_value()
+{
+	unsigned long cardID;
+
+  if ((_bitCount==24) || (_bitCount==26) || (_bitCount==32) || (_bitCount==34) || (_bitCount==8) || (_bitCount==4)) 	// bitCount for keypress=4 or 8, Wiegand 26=24 or 26, Wiegand 34=32 or 34
+  {
+    _cardTemp >>= 1;			// shift right 1 bit to get back the real value - interrupt done 1 left shift in advance
+    if (_bitCount>32)			// bit count more than 32 bits, shift high bits right to make adjustment
+    _cardTempHigh >>= 1;	
+
+    if (_bitCount==8)		// keypress wiegand with integrity
+    {
+      // 8-bit Wiegand keyboard data, high nibble is the "NOT" of low nibble
+      // eg if key 1 pressed, data=E1 in binary 11100001 , high nibble=1110 , low nibble = 0001 
+      char highNibble = (_cardTemp & 0xf0) >>4;
+      char lowNibble = (_cardTemp & 0x0f);
+      _wiegandType=_bitCount;					
+      _bitCount=0;
+      _cardTemp=0;
+      _cardTempHigh=0;
+      
+      if (lowNibble == (~highNibble & 0x0f))		// check if low nibble matches the "NOT" of high nibble.
+      {
+        _code = (int)translateEnterEscapeKeyPress(lowNibble);
+        return true;
+      }
+      else {
+        _bitCount=0;
+        _cardTemp=0;
+        _cardTempHigh=0;
+        return false;
+      }
+
+      // TODO: Handle validation failure case!
+    }
+    else if (4 == _bitCount) {
+      // 4-bit Wiegand codes have no data integrity check so we just
+      // read the LOW nibble.
+      _code = (int)translateEnterEscapeKeyPress(_cardTemp & 0x0000000F);
+
+      _wiegandType = _bitCount;
+      _bitCount = 0;
+      _cardTemp = 0;
+      _cardTempHigh = 0;
+
+      return true;
+    }
+    else		// wiegand 26 or wiegand 34
+    {
+      cardID = wiegand_get_card_id (&_cardTempHigh, &_cardTemp, _bitCount);
+      _wiegandType=_bitCount;
+      _bitCount=0;
+      _cardTemp=0;
+      _cardTempHigh=0;
+      _code=cardID;
+      return true;
+    }
+  }
+  else
+  {
+    // well time over 25 ms and bitCount !=8 , !=26, !=34 , must be noise or nothing then.
+    _bitCount=0;			
+    _cardTemp=0;
+    _cardTempHigh=0;
+    return false;
+  }	
+}
+
+static void lwiegand_cb(os_param_t param, uint8_t prio)
+{
+  (void) param;
+  (void) prio;
+
+  if (wiegand_timer_running)
+    os_timer_disarm(&wiegand_timer);
+
+  os_timer_arm(&wiegand_timer, 25, 0);
+}
+
+static void lwiegand_timer_done(void *param)
+{
+  lua_State *L = lua_getstate();
+
+  (void) param;
+
+  os_timer_disarm(&wiegand_timer);
+
+  if (wiegand_store_value()) {
+    if (wiegand_cb_ref != LUA_NOREF) {
+
+      lua_rawgeti(L, LUA_REGISTRYINDEX, wiegand_cb_ref);
+
+      lua_pushinteger(L, _code);
+      lua_pushinteger(L, _wiegandType);
+
+      lua_call(L, 2, 0);
+    }
+  }
+}
+
+static int lwiegand_close( lua_State* L)
+{
+  if(wiegand_cb_ref != LUA_NOREF) {
+    luaL_unref(L, LUA_REGISTRYINDEX, wiegand_cb_ref);
+    wiegand_cb_ref = LUA_NOREF;
+  }
+  return 0;
```

Comment:
I suspect that you ought to release the interrupt hook here.

Reply:
I was trying to figure out how to do that. AFAICT, the softuart module doesn't ever unregister its hook. The rotary module does, but using a lower level function. I can't figure out the the gpio module explicitly registers their hook - it seems it's just uses an even lower level function to create a fallback hook. Any tips on the recommended way to do it?

Reply:
If there isn't an easy way to do this, just make sure that nothing bad happens if the interrupt hook is called when the device is closed. Also, if the device is reopened, then don't reregister the hook (or maybe the hook registration code already takes care of it). Please add a comment to say what strategy you are following in the close function.